text/0000-glimmer-components.md

```diff
@@ -0,0 +1,835 @@
+oriented components.
+
+### Glimmer.js Compatibility
+
+One of the goals for future versions of Ember, post Ember Octane, will be to
+enable lighter-weight applications to be built using the framework. Breaking
+Ember apart into smaller, fully independent and optional pieces is the core idea
+behind the "install your way to Ember" goal, which will enable Ember to be used
+in more constrained environments that smaller frameworks such as React, Preact,
+Vue, and more excel in. It will also allow users who are size-conscious to adopt
+Ember incrementally, adding functionality when it is _needed_ rather than having
+all-or-nothing.
+
+This will take time though. Much progress has been made, but parts of Ember are
+still very monolithic. And while it isn't Ember, Glimmer.js is a _very_
+lightweight wrapper of the Glimmer VM that enables users to drop that weight and
+begin writing much more minimal apps today.
+
+Glimmer components aren't just based on the Glimmer.js component API - they are
+one and the same. They will be a shared package, which will be importable and
+usable by the users of both frameworks. Not only will users be able to write
+better components in Ember, those components will also be cross-compatible with
+Glimmer apps (assuming they don't use Ember specific functionality).
+
+It is important to note that while Glimmer components will be versioned
+independently from Glimmer and Ember ***they will abide by the Ember RFC process
+for any and all changes to user APIs***. The implementations for their component
+managers in Glimmer and Ember may change to keep them compatible, but they will
+not make major changes without first getting community input, and will be
+considered part of the public API of Ember.
+
+## Prior Art
+
+* [Glimmer.js](https://glimmerjs.com/) is the framework that this component API
+  is based on, and will be cross-compatible with.
+* [sparkles-component](https://github.com/rwjblue/sparkles-component) is an
+  an implementation of the Glimmer.js component API using Ember's component
+  managers. It is usable in Ember today.
+
+## Detailed design
+
+Glimmer components have the following API:
+
+```ts
+class GlimmerComponent<Args> {
+  constructor(injections: object, public args: Args);
+
+  didCreate(): void;
+  didUpdate(): void;
+
+  destroy(): void;
+}
+```
+
+In addition, two new element modifiers will fill in the gaps left by the
+`didInsertElement` and `willDestroyElement` lifecycle hooks:
+
+* `{{on-insert}}`
+* `{{on-destroy}}`
+
+Finally, Glimmer components will diverge from standard Ember components in one
+major way - untracked properties (non-computed, non-tracked, and not named args)
+will never update. This means that changes caused by `notifyPropertyChange` and
+`set` will not be picked up, unless they go through one of the above methods
+for tracking.
+
+### Constructor
+
+The constructor for Glimmer components receives two arguments: An injections
```

Comment:
does the GlimmerComponent's constructor assert that these arguments are passed, or are they optional?

Reply:
It should probably assert I think, it would be pretty bad if injections were not passed to super. Args is less bad, but I still think it would be pretty bad since it invalidates our invariant (args always === invocation args)